src/osd/scrubber/scrub_backend.cc

```diff
@@ -1866,80 +1840,50 @@ void ScrubBackend::scan_snaps(ScrubMap& smap)
-                                     const SnapSet& snapset)
+std::optional<snap_mapper_fix_t> ScrubBackend::scan_object_snaps(
+  const hobject_t& hoid,
+  const SnapSet& snapset,
+  snap_mapper_accessor_t& snaps_getter)
 {
   // check and if necessary fix snap_mapper
 
   auto p = snapset.clone_snaps.find(hoid.snap);
   if (p == snapset.clone_snaps.end()) {
     derr << __func__ << " no clone_snaps for " << hoid << " in " << snapset
          << dendl;
-    return;
+    return std::nullopt;
   }
   set<snapid_t> obj_snaps{p->second.begin(), p->second.end()};
 
   set<snapid_t> cur_snaps;
-  int r = m_pg.snap_mapper.get_snaps(hoid, &cur_snaps);
+  int r = snaps_getter.get_snaps(hoid, cur_snaps);
   if (r != 0 && r != -ENOENT) {
     derr << __func__ << ": get_snaps returned " << cpp_strerror(r) << dendl;
     ceph_abort();
   }
   if (r == -ENOENT || cur_snaps != obj_snaps) {
-    ObjectStore::Transaction t;
-    OSDriver::OSTransaction _t(m_pg.osdriver.get_transaction(&t));
-    if (r == 0) {
-      r = m_pg.snap_mapper.remove_oid(hoid, &_t);
-      if (r != 0) {
-        derr << __func__ << ": remove_oid returned " << cpp_strerror(r)
-             << dendl;
-        ceph_abort();
-      }
-      clog->error() << "osd." << m_pg_whoami
-                    << " found snap mapper error on pg " << m_pg_id << " oid "
-                    << hoid << " snaps in mapper: " << cur_snaps
-                    << ", oi: " << obj_snaps << "...repaired";
-    } else {
-      clog->error() << "osd." << m_pg_whoami
-                    << " found snap mapper error on pg " << m_pg_id << " oid "
-                    << hoid << " snaps missing in mapper"
-                    << ", should be: " << obj_snaps << " was " << cur_snaps
-                    << " r " << r << "...repaired";
-    }
-    m_pg.snap_mapper.add_oid(hoid, obj_snaps, &_t);
-
-    // wait for repair to apply to avoid confusing other bits of the system.
-    {
-      dout(15) << __func__ << " wait on repair!" << dendl;
-
-      ceph::condition_variable my_cond;
-      ceph::mutex my_lock = ceph::make_mutex("PG::_scan_snaps my_lock");
-      int e = 0;
-      bool done;  // note: initialized to 'false' by C_SafeCond
-
-      t.register_on_applied_sync(new C_SafeCond(my_lock, my_cond, &done, &e));
 
-      e = m_pg.osd->store->queue_transaction(m_pg.ch, std::move(t));
-      if (e != 0) {
-        derr << __func__ << ": queue_transaction got " << cpp_strerror(e)
-             << dendl;
-      } else {
-        std::unique_lock l{my_lock};
-        my_cond.wait(l, [&done] { return done; });
-        ceph_assert(m_pg.osd->store);
-      }
-      dout(15) << __func__ << " wait on repair - done" << dendl;
-    }
+    // add this object to the list of snapsets that needs fixing. Note
+    // that we also collect the existing (bogus) list, as legacy log lines show
+    // those
+    snap_mapper_op_t fixing_op =
+      (r == -ENOENT ? snap_mapper_op_t::add : snap_mapper_op_t::update);
```

Comment:
Is this distinction actually important?  Can we unconditionally clear and then set?

Reply:
I wanted to maintain the original logs emitted for each case (also - that's extra work with no good reason).